File: lab3/gr1sub1/KuharchukAlesia/src/pstat.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#define MAX_BUF 4096
#define HZ 100  // Частота системного таймера

typedef struct {
    int pid;
    char comm[256];
    char state;
    int ppid;
    unsigned long utime;
    unsigned long stime;
    long num_threads;
} proc_stat_t;

typedef struct {
    unsigned long vm_rss;
    unsigned long rss_anon;
    unsigned long rss_file;
    unsigned long voluntary_ctxt_switches;
    unsigned long nonvoluntary_ctxt_switches;
} proc_status_t;

typedef struct {
    unsigned long read_bytes;
    unsigned long write_bytes;
} proc_io_t;

int read_proc_file(const char* filename, char* buffer, size_t size) {
    FILE* file = fopen(filename, "r");
    if (!file) {
        return -1;
    }
    
    size_t bytes_read = fread(buffer, 1, size - 1, file);
    buffer[bytes_read] = '\0';
    fclose(file);
    
    return 0;
}

int parse_proc_stat(pid_t pid, proc_stat_t* stat) {
    char filename[64];
    char buffer[MAX_BUF];
    
    snprintf(filename, sizeof(filename), "/proc/%d/stat", pid);
    
    if (read_proc_file(filename, buffer, sizeof(buffer)) < 0) {
        return -1;
    }
    
    sscanf(buffer, 
           "%d %s %c %d %*d %*d %*d %*d %*u %*u %*u %*u %*u %lu %lu %*d %*d %*d %*d %*d %*d %*u %*u %ld",
           &stat->pid, stat->comm, &stat->state, &stat->ppid,
           &stat->utime, &stat->stime, &stat->num_threads);
    
    return 0;
}

int parse_proc_status(pid_t pid, proc_status_t* status) {
    char filename[64];
    char buffer[MAX_BUF];
    char line[256];
    
    snprintf(filename, sizeof(filename), "/proc/%d/status", pid);
    
    if (read_proc_file(filename, buffer, sizeof(buffer)) < 0) {
        return -1;
    }
    
    // Ищем нужные поля в файле status
    char* pos = buffer;
    while (*pos) {
        if (strncmp(pos, "VmRSS:", 6) == 0) {
            sscanf(pos + 6, "%lu", &status->vm_rss);
        } else if (strncmp(pos, "RssAnon:", 8) == 0) {
            sscanf(pos + 8, "%lu", &status->rss_anon);
        } else if (strncmp(pos, "RssFile:", 8) == 0) {
            sscanf(pos + 8, "%lu", &status->rss_file);
        } else if (strncmp(pos, "voluntary_ctxt_switches:", 24) == 0) {
            sscanf(pos + 24, "%lu", &status->voluntary_ctxt_switches);
        } else if (strncmp(pos, "nonvoluntary_ctxt_switches:", 27) == 0) {
            sscanf(pos + 27, "%lu", &status->nonvoluntary_ctxt_switches);
        }
        pos = strchr(pos, '\n');
        if (!pos) break;
        pos++;
    }
    
    return 0;
}

int parse_proc_io(pid_t pid, proc_io_t* io) {
    char filename[64];
    char buffer[MAX_BUF];
    char line[256];
    
    snprintf(filename, sizeof(filename), "/proc/%d/io", pid);
    
    if (read_proc_file(filename, buffer, sizeof(buffer)) < 0) {
        return -1;
    }
    
    char* pos = buffer;
    while (*pos) {
        if (strncmp(pos, "read_bytes:", 11) == 0) {
            sscanf(pos + 11, "%lu", &io->read_bytes);
        } else if (strncmp(pos, "write_bytes:", 12) == 0) {
            sscanf(pos + 12, "%lu", &io->write_bytes);
        }
        pos = strchr(pos, '\n');
        if (!pos) break;
        pos++;
    }
    
    return 0;
}

void format_memory_size(unsigned long kb, char* buffer, size_t size) {
    if (kb >= 1024) {
        snprintf(buffer, size, "%.2f MiB", kb / 1024.0);
    } else {
        snprintf(buffer, size, "%lu KiB", kb);
    }
}

void format_io_size(unsigned long bytes, char* buffer, size_t size) {
    if (bytes >= 1024 * 1024) {
        snprintf(buffer, size, "%.2f MiB", bytes / (1024.0 * 1024.0));
    } else if (bytes >= 1024) {
        snprintf(buffer, size, "%.2f KiB", bytes / 1024.0);
    } else {
        snprintf(buffer, size, "%lu bytes", bytes);
    }
}
int main(int argc, char* argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <pid>\n", argv[0]);
        return 1;
    }
    
    pid_t pid = atoi(argv[1]);
    
    proc_stat_t stat;
    proc_status_t status;
    proc_io_t io;
    
    // Читаем данные из /proc
    if (parse_proc_stat(pid, &stat) < 0) {
        fprintf(stderr, "Error reading /proc/%d/stat\n", pid);
        return 1;
    }
    
    if (parse_proc_status(pid, &status) < 0) {
        fprintf(stderr, "Error reading /proc/%d/status\n", pid);
        return 1;
    }
    
    if (parse_proc_io(pid, &io) < 0) {
        fprintf(stderr, "Warning: Cannot read /proc/%d/io\n", pid);
        memset(&io, 0, sizeof(io));
    }
    
    // Форматируем вывод
    char rss_mib[64], rss_anon_fmt[64], rss_file_fmt[64];
    char read_bytes_fmt[64], write_bytes_fmt[64];
    
    format_memory_size(status.vm_rss, rss_mib, sizeof(rss_mib));
    format_memory_size(status.rss_anon, rss_anon_fmt, sizeof(rss_anon_fmt));
    format_memory_size(status.rss_file, rss_file_fmt, sizeof(rss_file_fmt));
    format_io_size(io.read_bytes, read_bytes_fmt, sizeof(read_bytes_fmt));
    format_io_size(io.write_bytes, write_bytes_fmt, sizeof(write_bytes_fmt));
    
    double cpu_time_sec = (stat.utime + stat.stime) / (double)HZ;
    
    // Выводим сводку
    printf("Process %d (%s) statistics:\n", pid, stat.comm);
    printf("========================================\n");
    printf("Basic Information:\n");
    printf("  PPid: %d\n", stat.ppid);
    printf("  Threads: %ld\n", stat.num_threads);
    printf("  State: %c\n", stat.state);
    printf("  CPU time: %.2f sec (utime: %lu, stime: %lu)\n", 
           cpu_time_sec, stat.utime, stat.stime);
    
    printf("\nMemory Information:\n");
    printf("  VmRSS: %s\n", rss_mib);
    printf("  RssAnon: %s\n", rss_anon_fmt);
    printf("  RssFile: %s\n", rss_file_fmt);
    
    printf("\nContext Switches:\n");
    printf("  Voluntary: %lu\n", status.voluntary_ctxt_switches);
    printf("  Non-voluntary: %lu\n", status.nonvoluntary_ctxt_switches);
    
    printf("\nI/O Information:\n");
    printf("  Read bytes: %s\n", read_bytes_fmt);
    printf("  Write bytes: %s\n", write_bytes_fmt);
    
    return 0;
}

File: lab3/gr9sub1/TITKOVA_ANNA/src/cmd/pstat/pstat.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"syscall"
)

const version = "pstat-go-1.0"

type ProcStat struct {
	Pid        int
	Comm       string
	State      string
	Ppid       int
	UTime      uint64
	STime      uint64
	NumThreads int
	RSSPages   int64
}

type ProcStatus struct {
	VmRSS                    uint64
	VmSize                   uint64
	Threads                  uint64
	VoluntaryCtxtSwitches    uint64
	NonvoluntaryCtxtSwitches uint64
}

type ProcIO struct {
	RChar      uint64
	WChar      uint64
	ReadBytes  uint64
	WriteBytes uint64
}

type SmapsRollup struct {
	Rss   uint64
	Pss   uint64
	Anon  uint64
	File  uint64
	Shmem uint64
}

func formatBytes(bytes uint64) string {
	const (
		KB = 1024
		MB = KB * 1024
		GB = MB * 1024
	)

	switch {
	case bytes >= GB:
		return fmt.Sprintf("%.2f GiB", float64(bytes)/float64(GB))
	case bytes >= MB:
		return fmt.Sprintf("%.2f MiB", float64(bytes)/float64(MB))
	case bytes >= KB:
		return fmt.Sprintf("%.1f KiB", float64(bytes)/float64(KB))
	default:
		return fmt.Sprintf("%d B", bytes)
	}
}

func readProcFile(pid int, filename string) (string, error) {
	path := filepath.Join("/proc", strconv.Itoa(pid), filename)
	content, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	return string(content), nil
}

func parseProcStat(pid int) (*ProcStat, error) {
	content, err := readProcFile(pid, "stat")
	if err != nil {
		return nil, err
	}

	// Find the command name between parentheses
	start := strings.IndexByte(content, '(')
	if start == -1 {
		return nil, fmt.Errorf("invalid stat format")
	}
	end := strings.LastIndexByte(content, ')')
	if end == -1 {
		return nil, fmt.Errorf("invalid stat format")
	}

	// Extract command name and split the rest
	comm := content[start+1 : end]
	rest := strings.Fields(content[end+2:]) // Skip ") "

	// Parse fields (positions are 1-based in stat, but we have 0-based array)
	if len(rest) < 22 {
		return nil, fmt.Errorf("invalid stat format: not enough fields")
	}

	stat := &ProcStat{
		Pid:        pid,
		Comm:       comm,
		State:      rest[0],          // field 3
		Ppid:       atoi(rest[1]),    // field 4
		UTime:      atou(rest[11]),   // field 14
		STime:      atou(rest[12]),   // field 15
		NumThreads: atoi(rest[17]),   // field 20
		RSSPages:   atoi64(rest[21]), // field 24
	}

	return stat, nil
}

func parseProcStatus(pid int) (*ProcStatus, error) {
	content, err := readProcFile(pid, "status")
	if err != nil {
		return nil, err
	}

	status := &ProcStatus{}
	scanner := bufio.NewScanner(strings.NewReader(content))

	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.SplitN(line, ":", 2)
		if len(parts) != 2 {
			continue
		}

		key := strings.TrimSpace(parts[0])
		value := strings.TrimSpace(parts[1])

		switch key {
		case "VmRSS":
			status.VmRSS = parseMemoryValue(value)
		case "VmSize":
			status.VmSize = parseMemoryValue(value)
		case "Threads":
			status.Threads = atou(value)
		case "voluntary_ctxt_switches":
			status.VoluntaryCtxtSwitches = atou(value)
		case "nonvoluntary_ctxt_switches":
			status.NonvoluntaryCtxtSwitches = atou(value)
		}
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return status, nil
}

func parseProcIO(pid int) (*ProcIO, error) {
	content, err := readProcFile(pid, "io")
	if err != nil {
		return nil, err
	}

	io := &ProcIO{}
	scanner := bufio.NewScanner(strings.NewReader(content))

	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.SplitN(line, ":", 2)
		if len(parts) != 2 {
			continue
		}

		key := strings.TrimSpace(parts[0])
		value := strings.TrimSpace(parts[1])

		switch key {
		case "rchar":
			io.RChar = atou(value)
		case "wchar":
			io.WChar = atou(value)
		case "read_bytes":
			io.ReadBytes = atou(value)
		case "write_bytes":
			io.WriteBytes = atou(value)
		}
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return io, nil
}

func parseSmapsRollup(pid int) (*SmapsRollup, error) {
	content, err := readProcFile(pid, "smaps_rollup")
	if err != nil {
		return nil, err
	}

	smaps := &SmapsRollup{}
	scanner := bufio.NewScanner(strings.NewReader(content))

	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.SplitN(line, ":", 2)
		if len(parts) != 2 {
			continue
		}

		key := strings.TrimSpace(parts[0])
		value := strings.TrimSpace(parts[1])

		switch key {
		case "Rss":
			smaps.Rss = parseMemoryValue(value)
		case "Pss":
			smaps.Pss = parseMemoryValue(value)
		case "Anonymous":
			smaps.Anon = parseMemoryValue(value)
		case "File":
			smaps.File = parseMemoryValue(value)
		case "Shared_Hugetlb":
			smaps.Shmem = parseMemoryValue(value)
		}
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return smaps, nil
}

func parseMemoryValue(value string) uint64 {
	// Remove "kB" suffix and convert to bytes
	re := regexp.MustCompile(`(\d+)\s*kB`)
	matches := re.FindStringSubmatch(value)
	if len(matches) == 2 {
		return atou(matches[1]) * 1024
	}

	// If no "kB" suffix, try to parse as raw number
	fields := strings.Fields(value)
	if len(fields) > 0 {
		return atou(fields[0])
	}
	return 0
}

func atoi(s string) int {
	val, err := strconv.Atoi(s)
	if err != nil {
		return 0
	}
	return val
}

func atoi64(s string) int64 {
	val, err := strconv.ParseInt(s, 10, 64)
	if err != nil {
		return 0
	}
	return val
}

func atou(s string) uint64 {
	val, err := strconv.ParseUint(s, 10, 64)
	if err != nil {
		return 0
	}
	return val
}

func getSystemConstants() (clkTck uint64, pageSize uint64) {
	clkTck = 100 // Standard HZ value for most Linux systems
	pageSize = uint64(syscall.Getpagesize())
	return
}

func runCommand(cmd string) (string, error) {
	output, err := exec.Command("sh", "-c", cmd).CombinedOutput()
	if err != nil {
		return string(output), err
	}
	return string(output), nil
}

func runSystemComparisons(pid int) {
	commands := []struct {
		name string
		cmd  string
	}{
		{"ps", fmt.Sprintf("ps -p %d -o pid,ppid,comm,state,time,rss,vsz --no-headers", pid)},
		{"top", fmt.Sprintf("top -b -n1 -p %d 2>/dev/null | head -20", pid)},
		{"pidstat", fmt.Sprintf("pidstat -p %d 1 1 2>/dev/null || echo '(pidstat not available)'", pid)},
	}

	for _, cmd := range commands {
		fmt.Printf("\n=== %s ===\n", cmd.name)
		output, err := runCommand(cmd.cmd)
		if err != nil {
			fmt.Printf("Error: %v\n", err)
		} else {
			fmt.Print(output)
		}
	}
}

func printSummary(pid int, noCompare bool) {
	fmt.Printf("pstat-go v%s — quick snapshot for PID %d\n\n", version, pid)

	// Get system constants
	clkTck, pageSize := getSystemConstants()
	fmt.Printf("system: CLK_TCK=%d  PAGE_SIZE=%d bytes\n", clkTck, pageSize)

	// Parse proc files
	stat, err := parseProcStat(pid)
	if err != nil {
		fmt.Printf("Error reading stat: %v\n", err)
		return
	}

	status, err := parseProcStatus(pid)
	if err != nil {
		fmt.Printf("Error reading status: %v\n", err)
		return
	}

	io, err := parseProcIO(pid)
	if err != nil {
		fmt.Printf("Error reading io: %v\n", err)
	}

	smaps, err := parseSmapsRollup(pid)
	if err != nil {
		fmt.Printf("Note: smaps_rollup not available: %v\n", err)
	}

	// Calculate derived values
	cpuTimeSec := float64(stat.UTime+stat.STime) / float64(clkTck)
	rssBytes := uint64(stat.RSSPages) * pageSize

	// Print process summary
	fmt.Println("--- process summary ---")
	fmt.Printf("PID/Name : %d / %s\n", stat.Pid, stat.Comm)
	fmt.Printf("PPid     : %d    Threads: %d    State: %s\n", stat.Ppid, stat.NumThreads, stat.State)
	fmt.Printf("CPU(ticks): utime=%d stime=%d   CPU seconds=%d/%d = %.3fs\n",
		stat.UTime, stat.STime, stat.UTime+stat.STime, clkTck, cpuTimeSec)

	// Memory information
	fmt.Println("\n--- memory ---")
	fmt.Printf("Memory RSS : %d kB (%s)\n", status.VmRSS/1024, formatBytes(status.VmRSS))
	fmt.Printf("Stat RSS   : %d bytes (%s)\n", rssBytes, formatBytes(rssBytes))

	if status.VmSize > 0 {
		perc := float64(status.VmRSS) / float64(status.VmSize) * 100
		fmt.Printf("VmSize     : %d kB (%s)   RSS/VmSize = %.1f%%\n",
			status.VmSize/1024, formatBytes(status.VmSize), perc)
	}

	// Smaps information
	if smaps != nil {
		fmt.Println("\n--- smaps (rollup) ---")
		fmt.Printf("  Rss: %d kB (%s)\n", smaps.Rss/1024, formatBytes(smaps.Rss))
		fmt.Printf("  Anonymous: %d kB (%s)\n", smaps.Anon/1024, formatBytes(smaps.Anon))
		fmt.Printf("  File: %d kB (%s)\n", smaps.File/1024, formatBytes(smaps.File))
		fmt.Printf("  Shared: %d kB (%s)\n", smaps.Shmem/1024, formatBytes(smaps.Shmem))
	}

	// IO information
	if io != nil {
		fmt.Println("\n--- IO ---")
		fmt.Printf("rchar/wchar : %d / %d  (syscall bytes)\n", io.RChar, io.WChar)
		fmt.Printf("read_bytes/write_bytes: %d / %d  (%s / %s)\n",
			io.ReadBytes, io.WriteBytes, formatBytes(io.ReadBytes), formatBytes(io.WriteBytes))
	}

	// Context switches
	fmt.Println("\n--- context switches ---")
	fmt.Printf("voluntary   : %d\n", status.VoluntaryCtxtSwitches)
	fmt.Printf("non-voluntary: %d\n", status.NonvoluntaryCtxtSwitches)

	// System tools comparison
	if !noCompare {
		fmt.Println("\n--- comparison with system tools ---")
		runSystemComparisons(pid)
	}
}

func main() {
	var noCompare bool
	flag.BoolVar(&noCompare, "no-compare", false, "don't run system tool comparisons")
	flag.Usage = func() {
		fmt.Fprintf(flag.CommandLine.Output(), "Usage: %s [flags] <pid>\n", os.Args[0])
		fmt.Fprintf(flag.CommandLine.Output(), "Flags:\n")
		flag.PrintDefaults()
	}
	flag.Parse()

	if flag.NArg() != 1 {
		flag.Usage()
		os.Exit(1)
	}

	pid, err := strconv.Atoi(flag.Arg(0))
	if err != nil {
		fmt.Printf("Invalid PID: %s\n", flag.Arg(0))
		os.Exit(1)
	}

	printSummary(pid, noCompare)
}
